src/main/java/uk/gov/hmcts/reform/et/syaapi/service/CaseDocumentService.java

```
@@ -0,0 +1,211 @@
+     *                           Case Document API
+     * @param caseDocApiUrl      the URL to call the Case Document API
+     */
+    public CaseDocumentService(RestTemplate restTemplate,
+                               AuthTokenGenerator authTokenGenerator,
+                               @Value("${case_document_am.url}/cases/documents}")
+                                   String caseDocApiUrl) {
+        this.restTemplate = restTemplate;
+        this.authTokenGenerator = authTokenGenerator;
+        this.caseDocApiUrl = caseDocApiUrl;
+    }
+
+    /**
+     * Given a file to upload, this will upload the file to the CCD document API and give back a unique ULR to access
+     * the uploaded file.
+     *
+     * @param authToken  the caller's bearer token used to verify the caller
+     * @param caseTypeId the area the file belongs to e.g. ET_EnglandWales
+     * @param file       the file to be uploaded
+     * @return the URL of the document we have just uploaded
+     * @throws CaseDocumentException if a problem occurs whilst uploading the document via API
+     */
+    public URI uploadDocument(String authToken, String caseTypeId, MultipartFile file) throws CaseDocumentException {
+        DocumentUploadResponse response = attemptWithRetriesToUploadDocumentToCaseDocumentApi(
+            0, authToken, caseTypeId, file);
+
+        CaseDocument caseDocument = validateResponse(
+            Objects.requireNonNull(response), file.getOriginalFilename());
+
+        return getUriFromFile(caseDocument);
+    }
+
+    private URI getUriFromFile(CaseDocument caseDocument) {
+        if (caseDocument.getLinks() != null
+            && caseDocument.getLinks().get("self") != null
+            && caseDocument.getLinks().get("self").get("href") != null) {
+            return URI.create(caseDocument.getLinks().get("self").get("href"));
+        }
+        return URI.create("");
+    }
```

Worth just switching this around so if any are null, then return the empty.

But I'm also surprised how we can accept a blank URL being returned?  Would have thought that should throw an exception instead?  Only that what use would uploading a file to the service be if you can't get a handle on it?

---

Uploaded changes, also the blank url was caught by regex validation, but updated to throw error directly from this snippet